src/mender-auth/api/auth/embedded_auth.cpp

```diff
@@ -0,0 +1,68 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <mender-auth/api/auth.hpp>
+
+#include <common/log.hpp>
+
+namespace mender {
+namespace auth {
+namespace api {
+namespace auth {
+
+namespace log = mender::common::log;
+
+error::Error AuthenticatorHttp::StartWatchingTokenSignal() {
+	// There is no signal when embedding the authentication. We get the data straight from the
+	// HTTP response.
+	return error::NoError;
+}
+
+error::Error AuthenticatorHttp::GetJwtToken() {
+	loop_.Post([this]() {
+		HandleReceivedToken(common::StringPair {token_, server_url_}, NoTokenAction::RequestNew);
+	});
+	return error::NoError;
+}
+
+void AuthenticatorHttp::FetchJwtTokenHandler(APIResponse resp) {
+	if (resp) {
+		token_ = resp.value().token;
+		server_url_ = resp.value().server_url;
+
+		log::Info("Successfully received new authorization data");
+	} else {
+		token_.clear();
+		server_url_.clear();
+
+		log::Error("Failed to fetch new token: " + resp.error().String());
+	}
+
+	PostPendingActions(mender::api::auth::AuthData {token_, server_url_});
```

Comment:
Reasoning out loud and for me to understand,

Here there is a key difference between D-Bus and embedded: in D-Bus a fetch does `PostPendingActions` on error or return `false` from D-Bus, while the success case will eventually be called via the signal callback. Here we have one less level of async :+1:

Reply:
Exactly. Even the asking is async in the DBus case, while the actual reply (via signal) is separate, so we have two levels of async calls. Here we have only one, because we do the HTTP call directly.